[v24.2.x] storage: always schedule adjacent segment compaction

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
Backport of PR https://github.com/redpanda-data/redpanda/pull/24874

Also pulls in https://github.com/redpanda-data/redpanda/pull/24880 which is a dependency of this change
Fixes https://github.com/redpanda-data/redpanda/issues/24957

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [x] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Redpanda will now schedule local segment merges of compacted topics, even when windowed compaction has occurred in a given housekeeping round. This ensures progress in reducing segment count in compacted topics with high produce traffic.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
storage_e2e_test: add log messages indicating test progress

The test itself has many steps that make it hard to follow what has
happened. Adds some log messages to indicate different steps in the
test.

(cherry picked from commit 67689face63ca96b47e3ec56a1b7c9ae0ea411d3)

andrwng committed:
model: add bounded_offset_interval::optional()

Adds a static constructor to return an optional interval, which per the
class comments, should represent an empty interval.

I intend on using this for some upcoming bounds checks.

(cherry picked from commit 8514bfd917a4a4ee60cf8e6b1fc0a427a176f723)

andrwng committed:
storage: fix bounds check in offset range size method

The methods to find offset range size currently use segment bounds to
determine whether a given offset is available to be queried. This
doesn't account for the case when the segment set contains offsets that
don't fall in the log's offset range (e.g. follow a delete records
request that trims mid-segment).

This commit adds appropriate bounds checks to both methods.

With an upcoming change to merge compact after windowed compaction,
test_offset_range_size2_compacted would fail because it would prefix
truncate mid-segment following a merge compaction, and then trip over
this, hitting an unexpected exception when creating a reader:

```
std::runtime_error: Reader cannot read before start of the log 0 < 887
```

(cherry picked from commit 69e4666759eb90283530ce156e89ff784708faab)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62227-->
<!--7a787a4a-18aa-4828-adbe-8d6eb8a71bab-->
<!--980a1cdc-f0fb-42a3-b31c-c2b1e8f4a11e-->
<!--title-->
## Retry command for [Build#62227](https://buildkite.com/redpanda/redpanda/builds/62227)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/idempotency_test.py::IdempotencyWriteCachingTest.test_idempotent_producers_write_caching
tests/rptest/tests/e2e_iam_role_test.py::ShortLivedCredentialsTests.test_short_lived_credentials
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--476a9c35-4581-4b98-a083-d508bc459013-->
<!--TEST RESULTS-->
<!--c8f66be1-d696-450a-8d6b-610ec437e4bd-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62227">build#62227</a></summary>


| test_id                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit                                                                  | unit        | https://buildkite.com/redpanda/redpanda/builds/62227#01953a3c-0647-4239-96ec-55095711d80f | FLAKY         | 1/2      |
| rptest.tests.consumer_group_balancing_test.ConsumerGroupBalancingTest.test_coordinator_nodes_balance | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62227#01953a85-bc5d-450e-a612-0f339898d367 | FLAKY         | 5/6      |
| rptest.tests.e2e_iam_role_test.ShortLivedCredentialsTests.test_short_lived_credentials               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62227#01953a85-bc5d-450e-a612-0f339898d367 | FAIL          | 0/1      |
| rptest.tests.idempotency_test.IdempotencyWriteCachingTest.test_idempotent_producers_write_caching    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62227#01953a85-6923-4643-9e27-b9ea7ffacbb9 | FAIL          | 0/1      |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62227#01953a85-6922-48ac-8ab8-50afd8f00d91 | FLAKY         | 5/6      |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/62227#01953a3c-0647-4239-96ec-55095711d80f | FAIL          | 0/2      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65940">build#65940</a></summary>


| test_class                       | test_method                                       | test_arguments                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      |
|:---------------------------------|:--------------------------------------------------|:-------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest     | test_cloud_storage_with_partition_moves           | {"cleanup_policy": "compact,delete"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65940#0196cb95-e8a4-4434-be12-be841a8e8876 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 0. The test should FAIL    |
| CompactionRecoveryTest           | test_index_recovery                               |                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65940#0196cb8d-52ea-4b2c-9af4-8e22e2ab9d9d | FLAKY         | 8/21     | upstream reliability is '100.0'. current run reliability is '38.095238095238095'. drift is 61.90476 and the allowed drift is set to 0. The test should FAIL |
| EnterpriseFeaturesTest           | test_license_violation                            | {"feature": 5, "with_license": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65940#0196cb95-e8a3-4c75-aca7-1eb9fce8217a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 0. The test should FAIL    |
| PartitionReassignmentsTest       | test_add_partitions_with_inprogress_reassignments |                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65940#0196cb95-e8a3-4c75-aca7-1eb9fce8217a | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 0. The test should FAIL  |
| storage_e2e_single_thread_rpunit | storage_e2e_single_thread_rpunit                  |                                      | unit        | https://buildkite.com/redpanda/redpanda/builds/65940#0196cb4d-d5fb-4a77-80e8-af4e760a2c26 | FAIL          | 0/2      |                                                                                                                                                             |
| storage_e2e_single_thread_rpunit | storage_e2e_single_thread_rpunit                  |                                      | unit        | https://buildkite.com/redpanda/redpanda/builds/65940#0196cb4d-d5fd-48cc-904c-e6c29bf3c21e | FAIL          | 0/2      |                                                                                                                                                             |
</details>
<!--/body-->

andrwng committed:
storage: always schedule adjacent segment compaction

CONFLICT:
- this branch had code that takes a range instead of just the compaction
  config for compact_adjacent_segments()

We previously fell back on adjacent segment compaction only if there was
no new data to compact. In some situations, we've seen the rate of
incoming data outpace the compaction interval, causing segments to pile
up without ever being merged.

This change tweaks the logic to always run adjacent segment compaction
after running sliding window compaction.

Along the way, a couple tests needed to be tweaked to handle the fact
that housekeeping now may merge segments.

(cherry picked from commit 08d0433a2a93990a6c14eb69fc33c670ad2e64d0)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65940-->
<!--5459b014-4699-4398-b330-584b7bc7daf8-->
<!--fabb28db-09cf-48e5-9a9e-6dc4b9514a22-->
<!--a6f33261-846a-49f1-8632-ebdc327950a3-->
<!--f9fa555f-edaf-489d-b93e-c2154e6ae34d-->
<!--1b48b452-61f2-4070-86f0-e0d4e328cc00-->
<!--2198c4c6-2829-4da6-8c30-166a6f62e65a-->
<!--title-->
## Retry command for [Build#65940](https://buildkite.com/redpanda/redpanda/builds/65940)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/compaction_recovery_test.py::CompactionRecoveryTest.test_index_recovery
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_license_violation@{"feature":5,"with_license":true}
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves@{"cleanup_policy":"compact,delete"}
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_transfer_controller_leadership
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_unavailable_nodes
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/compaction_recovery_test.py::CompactionRecoveryTest.test_index_recovery
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_license_violation@{"feature":5,"with_license":true}
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves@{"cleanup_policy":"compact,delete"}
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_transfer_controller_leadership
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_unavailable_nodes
